### ISSUE TITLE
When Publishing Command - MetadataKeyNotFoundException: Specified argument was out of the range 'previous_source_ids'

### ISSUE THREAD

**Issue by laalkep** [rasmus/EventFlow#889]
I get the below exception whenever I try to publish a command related to one particular AggregateRoot.  The aggregate has snapshots and it seems to be that that fails.  I deleted the snapshots for that aggregate and it now works.

I have recently updated from version 0.81.4483 to 0.82.4659

I can work around it by deleting the old snapshots but I'm not sure that is a good idea.

Thanks.

```
MetadataKeyNotFoundException: Specified argument was out of the range of valid values. (Parameter 'Could not find metadata key 'previous_source_ids'')
EventFlow.Core.MetadataContainer.GetMetadataValue<T>(string key, Func<string, T> converter)
EventFlow.Snapshots.SnapshotMetadata.get_PreviousSourceIds()
EventFlow.Snapshots.SnapshotAggregateRoot<TAggregate, TIdentity, TSnapshot>.LoadAsync(IEventStore eventStore, ISnapshotStore snapshotStore, CancellationToken cancellationToken)
EventFlow.Aggregates.AggregateStore.LoadAsync<TAggregate, TIdentity>(TIdentity id, CancellationToken cancellationToken)
EventFlow.Aggregates.AggregateStore+<>c__DisplayClass13_0<TAggregate, TIdentity, TExecutionResult>+<<UpdateAsync>b__0>d.MoveNext()
EventFlow.Core.TransientFaultHandler<TRetryStrategy>.TryAsync<T>(Func<CancellationToken, Task<T>> action, Label label, CancellationToken cancellationToken)
EventFlow.Aggregates.AggregateStore.UpdateAsync<TAggregate, TIdentity, TExecutionResult>(TIdentity id, ISourceId sourceId, Func<TAggregate, CancellationToken, Task<TExecutionResult>> updateAggregate, CancellationToken cancellationToken)
EventFlow.CommandBus.ExecuteCommandAsync<TAggregate, TIdentity, TResult>(ICommand<TAggregate, TIdentity, TResult> command, CancellationToken cancellationToken)
EventFlow.CommandBus.PublishAsync<TAggregate, TIdentity, TResult>(ICommand<TAggregate, TIdentity, TResult> command, CancellationToken cancellationToken)
SuperFierce.UserInterface.Pages.Settings.SuperAverage.DeleteModel.OnPostAsync(string id, CancellationToken token) in Delete.cshtml.cs
+
            var result = await bus.PublishAsync(command, token)
```

**Comment by rasmus** [comment]
Hi @kepboy can you create a test that reproduces the problem?

**Comment by laalkep** [comment]
I've created a sample project that reproduces the error, not sure how an automated test could be constructed for this one.

To reproduce:
1. Unzip the project and open (you'll need visual studio and LocalDb)
2. Update database using Package-Manager
3. Run the simple command line program
4. It should create some events and at least 1 snapshot, you can check in the database
3. Upgrade the project to nuget package 0.82.4659
4. Attempt to run again -> exception occurs

Hopefully that is enough
Kep
[ev-prob.zip](https://github.com/eventflow/EventFlow/files/7063373/ev-prob.zip)

**Comment by kepboy** [comment]
Any ideas?

**Comment by laalkep** [comment]
I wonder if there should be a "TryGetMetadataValue" method in [MetadataContainer](https://github.com/eventflow/EventFlow/blob/develop-v1/Source/EventFlow/Core/MetadataContainer.cs)

And [SnapshotMetadata.cs](https://github.com/eventflow/EventFlow/blob/develop-v1/Source/EventFlow/Snapshots/SnapshotMetadata) should use that method, or catch the exception in the event the metadata does not contain that key.  Presumably that would result in an empty collection.

**Comment by rasmus** [comment]
The `TryGetMetadataValue` is better.

> Handle common conditions without throwing exceptions

https://docs.microsoft.com/en-us/dotnet/standard/exceptions/best-practices-for-exceptions

**Comment by rasmus** [comment]
I'll have a look

**Comment by rasmus** [comment]
Hi @kepboy I just did a quit spike on the bug. Haven't looked into the code or the tests yet. However, at a glance it seems like the PR might fix the issue. You are right, the `SnapshotMetadata`  should have used the existing `TryGetValue` instead.

**Comment by rasmus** [comment]
@kepboy released in 0.83.4713
